### PR TITLE
Refactor build-number/build-name logic.

### PR DESCRIPTION
**Pull request by kangwang1988**
This PR aims at several things:
1. Use pub_semver to check a version in pubspec.yaml meets the requirements specified in https://semver.org/.
2. Don't limit build-number/build-name as a fixed format. Instead, validate it according to the target(ios/android).
3. Make sure that build-number/build-name are always validated no matter it's specified by the `flutter command` or version in pubspec.yaml.

This will also fix:
https://github.com/flutter/flutter/issues/27589

### CODE REVIEW

File: packages/flutter_tools/test/build_info_test.dart [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter_tools/test/build_info_test.dart]: Should be 2019 if this is a new file.

**Review by gspencergoog — CHANGES_REQUESTED** [review]
Thanks for the this!

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
Oh! Sorry, @kangwang1988, I thought you still needed someone to merge your changes for you, but I see you're a project member now.  I'll let you merge your own from now on...

**Comment by kangwang1988** [comment]
@gspencergoog 
Thank you~

**Pull request merged**